core/quivr_core/rag/quivr_rag_langgraph.py

```diff
@@ -0,0 +1,930 @@
+        max_context_tokens: int | None = None,
+    ) -> Tuple[Dict[str, Any], List[Document] | None]:
+        MAX_ITERATION = 100
+        SECURITY_FACTOR = 0.85
+        iteration = 0
+
+        msg = prompt.format(**inputs)
+        n = self.llm_endpoint.count_tokens(msg)
+        reduced_inputs = deepcopy(inputs)
+
+        max_context_tokens = (
+            max_context_tokens
+            if max_context_tokens
+            else self.retrieval_config.llm_config.max_context_tokens
+        )
+
+        while n > max_context_tokens * SECURITY_FACTOR:
+            chat_history = (
+                reduced_inputs["chat_history"]
+                if "chat_history" in reduced_inputs
+                else []
+            )
+
+            if len(chat_history) > 0:
+                reduced_inputs["chat_history"] = chat_history[2:]
+            elif docs and len(docs) > 1:
+                docs = docs[:-1]
+            else:
+                logging.warning(
+                    f"Not enough context to reduce. The context length is {n} "
+                    f"which is greater than the max context tokens of {max_context_tokens}"
+                )
+                break
+
+            if docs and "context" in reduced_inputs:
+                reduced_inputs["context"] = combine_documents(docs)
+
+            msg = prompt.format(**reduced_inputs)
+            n = self.llm_endpoint.count_tokens(msg)
+
+            iteration += 1
+            if iteration > MAX_ITERATION:
+                logging.warning(
+                    f"Attained the maximum number of iterations ({MAX_ITERATION})"
+                )
+                break
+
+        return reduced_inputs, docs
+
+    def bind_tools_to_llm(self, node_name: str):
+        if self.llm_endpoint.supports_func_calling():
+            tools = self.retrieval_config.workflow_config.get_node_tools(node_name)
+            if tools:  # Only bind tools if there are any available
+                return self.llm_endpoint._llm.bind_tools(tools, tool_choice="any")
+        return self.llm_endpoint._llm
+
+    def generate_rag(self, state: AgentState) -> AgentState:
+        docs: List[Document] | None = state["docs"]
+        final_inputs = self._build_rag_prompt_inputs(state, docs)
+
+        reduced_inputs, docs = self.reduce_rag_context(
+            final_inputs, custom_prompts.RAG_ANSWER_PROMPT, docs
+        )
+
+        msg = custom_prompts.RAG_ANSWER_PROMPT.format(**reduced_inputs)
+        llm = self.bind_tools_to_llm(self.generate_rag.__name__)
+        response = llm.invoke(msg)
+
+        return {**state, "messages": [response], "docs": docs if docs else []}
+
+    def generate_chat_llm(self, state: AgentState) -> AgentState:
+        """
+        Generate answer
+
+        Args:
+            state (messages): The current state
+
+        Returns:
+            dict: The updated state with re-phrased question
+        """
+        messages = state["messages"]
```

Comment:
We can indeed use a BaseModel, but we should first remove and remaining dependency on pydantic V1, otherwise I keep gettign errors... I'll put this into an issue as well